File: 0289 Game of Life/0289.cpp
class Solution {
public:
    void nextState(vector<vector<int>>& board, int x, int y){
        int m = board.size();
        int n = board[0].size();

        int live = 0;

        for (int i = max(x-1, 0); i <= min(x+1, m-1); i++){
            for (int j = max(y-1, 0); j <= min(y+1, n-1); j++){
                if (!(x == i && y == j) && (board[i][j] == 1 || board[i][j] == 3)){
                    live++;
                }
            }
        }

        if (board[x][y] == 1 && live != 2 && live != 3){
            board[x][y] = 3;
        } else if (board[x][y] == 0 && live == 3){
            board[x][y] = 2;
        }
    }

    void cleanUp(vector<vector<int>>& board){
        int m = board.size();
        int n = board[0].size();

        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                if (board[i][j] == 2){
                    board[i][j] = 1;
                } else if (board[i][j] == 3){
                    board[i][j] = 0;
                }
            }
        }
    }


    void gameOfLife(vector<vector<int>>& board) {
        int m = board.size();
        int n = board[0].size();

        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                nextState(board, i, j);
            }
        }

        cleanUp(board);
    }
};

File: 0289 Game of Life/0289.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        m, n = len(board), len(board[0])

        def next_state(x, y):
            live = 0
            
            for i in range(max(x-1, 0), min(x+1, m-1) + 1):
                for j in range(max(y-1, 0), min(y+1, n-1) + 1):
                    if not (x == i and y == j) and (board[i][j] == 1 or board[i][j] == 3):
                        live += 1

            if board[x][y] == 1 and not (live == 2 or live == 3):
                board[x][y] = 3
            elif board[x][y] == 0 and live == 3:
                board[x][y] = 2

        for i in range(m):
            for j in range(n):
                next_state(i, j)

        for i in range(m):
            for j in range(n):
                if board[i][j] == 2:
                    board[i][j] = 1
                elif board[i][j] == 3:
                    board[i][j] = 0
